internal/tasks/config_storage.go

```diff
@@ -0,0 +1,150 @@
+package tasks
+
+import (
+	"context"
+	"crypto/md5" //nolint: gosec
+	"fmt"
+
+	"github.com/google/uuid"
+	"github.com/valkey-io/valkey-go"
+)
+
+type ConfigStoreKey[T any] struct {
+	Data []T
+}
+
+type TemplateVersionKey struct {
+	OrgID           uuid.UUID
+	Fleet           string
+	TemplateVersion string
+}
+
+func (k *TemplateVersionKey) ComposeKey() string {
+	return fmt.Sprintf("v1/%s/%s/%s/", k.OrgID, k.Fleet, k.TemplateVersion)
+}
+
+type ConfigStorageRepositoryUrlKey struct {
+	OrgID           uuid.UUID
+	Fleet           string
+	TemplateVersion string
+	Repository      string
+}
+
+func (k *ConfigStorageRepositoryUrlKey) ComposeKey() string {
+	return fmt.Sprintf("v1/%s/%s/%s/repo/%s", k.OrgID, k.Fleet, k.TemplateVersion, k.Repository)
+}
+
+type ConfigStorageGitRevisionKey struct {
+	OrgID           uuid.UUID
+	Fleet           string
+	TemplateVersion string
+	Repository      string
+	TargetRevision  string
+}
+
+func (k *ConfigStorageGitRevisionKey) ComposeKey() string {
+	return fmt.Sprintf("v1/%s/%s/%s/git/%s/%s", k.OrgID, k.Fleet, k.TemplateVersion, k.Repository, k.TargetRevision)
+}
+
+type ConfigStorageK8sSecretKey struct {
+	OrgID           uuid.UUID
+	Fleet           string
+	TemplateVersion string
+	Namespace       string
+	Name            string
+}
+
+func (k *ConfigStorageK8sSecretKey) ComposeKey() string {
+	return fmt.Sprintf("v1/%s/%s/%s/k8s/%s/%s", k.OrgID, k.Fleet, k.TemplateVersion, k.Namespace, k.Name)
+}
+
+type ConfigStorageHttpKey struct {
+	OrgID           uuid.UUID
+	Fleet           string
+	TemplateVersion string
+	URL             string
+}
+
+func (k *ConfigStorageHttpKey) ComposeKey() string {
+	md5sum := md5.Sum([]byte(k.URL)) //nolint: gosec
+	return fmt.Sprintf("v1/%s/%s/%s/http/%x", k.OrgID, k.Fleet, k.TemplateVersion, md5sum)
+}
+
+type ConfigStorage interface {
+	Close()
+	StoreIfNotExistsAndFetch(ctx context.Context, key string, value []byte) ([]byte, error)
+	StoreIfNotExists(ctx context.Context, key string, value []byte) error
```

Comment:
Could we make evaluations outside the methods and then use a simplified interface WDYT?

```go
type ConfigStorage interface {
	Close()
	Store(ctx context.Context, key string, value []byte) error
	Fetch(ctx context.Context, key string) ([]byte, error)                      
	Delete(ctx context.Context, key string) error
	DeleteByPrefix(ctx context.Context, prefix string) error 
	[..]
```

Reply:
What do you mean by "evaluations"?

Reply:
it seems that the methods do more than one thing. where you can have a simple interface and `Fetch()  then based on evaluation of result `Store()`. Just as an example.

Reply:
This isn't blocking just an observation